cassandra/pool.py

```diff
@@ -360,18 +363,45 @@ def __init__(self, host, host_distance, session):
             return
 
         log.debug("Initializing connection for host %s", self.host)
-        self._connection = session.cluster.connection_factory(host.endpoint)
+        first_connection = session.cluster.connection_factory(host.endpoint)
+        log.debug("first connection created for shard_id=%i", first_connection.shard_id)
+        self._connections[first_connection.shard_id] = first_connection
         self._keyspace = session.keyspace
+
         if self._keyspace:
-            self._connection.set_keyspace_blocking(self._keyspace)
+            first_connection.set_keyspace_blocking(self._keyspace)
+
+        if first_connection.sharding_info:
+            self.host.sharding_info = weakref.proxy(first_connection.sharding_info)
+            for _ in range(first_connection.sharding_info.shards_count * 2):
```

Comment:
can you explain the * 2 logic? I must miss some background which fails my understanding...

Reply:
the thing is there's no guarantee we'll get shards opened in a nicely sequence, hence on each iteration we might get a shared we already got, hence we need to continue this loop at least twice.
even that number might not be enough and would need to be enlarged to avoid failures to open all the shards. 

I'll add a comment to explain why we need this "magic" numbered loop.